### PR TITLE
Khalid test

### PR DESCRIPTION
**Pull request by KhalidTheeb**
Base
 Test using multiple words

Soft Hyphens
 Tests need to use browser.set_parameters()
 Test >2 lines where some soft hyphens don't split

### CODE REVIEW

**Review by pavpanchekha — CHANGES_REQUESTED** [review]

File: tests/chapter3-base-tests.md [hunk 1]
```diff
@@ -79,6 +81,10 @@ rather a mock font that has faked metrics.
     >>> test_layout("<big><big><i>abc</i></big>def</big>") #doctest: +NORMALIZE_WHITESPACE
     [(13.0, 22.5, 'abc', Font size=24 weight=normal slant=italic style=None), 
      (109.0, 25.5, 'def', Font size=20 weight=normal slant=roman style=None)]
+     
+    >>> test_layout("hello world")
+    [(13.0, 22.5, 'hello', Font size=16 weight=normal slant=roman style=None),
+	(13.0, 42.5, 'world', Font size=16 weight=normal slant=roman style=None)]
```
Comment (pavpanchekha) [inline tests/chapter3-base-tests.md:87]: Whitespace is wrong here, tests will not pass. You need to either:
- Put the list on one line (don't do this)
- Change `NORMALIZE_WHITESPACE` (don't do this either)
- Use `wbemocks.print_list` (do this one)

File: tests/chapter3-exercise-soft-hyphens-tests.md [hunk 2]
```diff
@@ -31,16 +43,21 @@ If the word fits without splitting then no literal hyphens are present.
...
+    >>> browser.set_parameters(WIDTH=122)
     >>> test_layout("multi\N{soft hyphen}word\N{soft hyphen}split") #doctest: +NORMALIZE_WHITESPACE
     [(13.0, 21.0, 'multi-', Font size=16 weight=normal slant=roman style=None), 
      (13.0, 41.0, 'word-', Font size=16 weight=normal slant=roman style=None),
      (13.0, 61.0, 'split', Font size=16 weight=normal slant=roman style=None)]
+     
+    >>> test_layout("multi\N{soft hyphen}word\N{soft hyphen}spl\N{soft hyphen}it") #doctest: +NORMALIZE_WHITESPACE
+    [(13.0, 21.0, 'multi-', Font size=16 weight=normal slant=roman style=None), 
+     (13.0, 41.0, 'word-', Font size=16 weight=normal slant=roman style=None),
+     (13.0, 61.0, 'split', Font size=16 weight=normal slant=roman style=None)]
```
Comment (pavpanchekha) [inline tests/chapter3-exercise-soft-hyphens-tests.md:63]: This one is good.

File: tests/chapter4-base-tests.md [hunk 3]
```diff
@@ -89,6 +89,19 @@ Attributes can be set on tags:
...
+
+	>>> test_parse("<i>A</i>B")
+	 <html>
+	   <body>
+		 <i>
+		   'A'
+		 </i>
+		 'B'
+	   </body>
+	 </html>
```
Comment (pavpanchekha) [inline tests/chapter4-base-tests.md:103]: Our `print_tree` doesn't print close tags.

File: tests/chapter4-exercise-comments-tests.md [hunk 4]
```diff
@@ -92,3 +98,25 @@ There are some edge cases to take care of.
      <html>
        <body>
          'thing'
+		 
+		 
+Test case for handling text outside of comments		 
+	>>> test_parse("This is normal text --> following text should also be normal")
```
Comment (pavpanchekha) [inline tests/chapter4-exercise-comments-tests.md:104]: Indentation issue

File: tests/chapter4-exercise-comments-tests.md [hunk 5]
```diff
@@ -92,3 +98,25 @@ There are some edge cases to take care of.
...
+     <html>
+       <body>
+         'This is normal text --> following text should also be normal'
+	 
+	 
+Test combinations "nested" cases
```
Comment (pavpanchekha) [inline tests/chapter4-exercise-comments-tests.md:110]: Add blank line between text & code blocks

File: tests/chapter4-exercise-comments-tests.md [hunk 6]
```diff
@@ -92,3 +98,25 @@ There are some edge cases to take care of.
...
+	>>> test_parse("This is normal text<!-- <!-- --> following text should also be normal")
+     <html>
+       <body>
+         'This is normal text following text should also be normal'
```
Comment (pavpanchekha) [inline tests/chapter4-exercise-comments-tests.md:114]: I think this should be two text nodes:
```
<html>
  <body>
    'This is normal text'
    ' following ...'
```

File: tests/chapter4-exercise-comments-tests.md [hunk 7]
```diff
@@ -92,3 +98,25 @@ There are some edge cases to take care of.
...
+
+	 
+ Test the case of <!-- !--> (should be one comment)
+ 	>>> test_parse("This is normal text<!-- !--> following text should also be normal")
+     <html>
+       <body>
+         'This is normal text following text should also be normal'
```
Comment (pavpanchekha) [inline tests/chapter4-exercise-comments-tests.md:121]: Same

File: tests/chapter4-exercise-paragraphs-tests.md [hunk 8]
```diff
@@ -1,14 +1,19 @@
...
+`<p>hello<p>world</p>` results in two sibling paragraphs instead of
+one paragraph inside another; real browsers do this too. 
+
+Make sure to handle cases like `<p><b><p>`. In this case, all tags
+inside the first paragraph should be closed and then re-opened in the
+second paragraph.
+
+
 Description
 ------------
```
Comment (pavpanchekha) [inline tests/chapter4-exercise-paragraphs-tests.md:15]: Don't need this now

File: tests/chapter4-exercise-paragraphs-tests.md [hunk 9]
```diff
@@ -48,3 +53,25 @@ Any tags that are open when encountering the second paragraph should be closed
          <p>
            <b>
              'world'
+			 
+			 
+Test when <p> tags have different attributes from each other			 
+	>>> test_parse("<p class='first-paragraph'>This has a class</p><p id='second-paragraph'>This has an ID</p>")
```
Comment (pavpanchekha) [inline tests/chapter4-exercise-paragraphs-tests.md:59]: I don't think this exercise triggers the "nested `<p>` tags" thing literally at all, since these `<p>` tags aren't nested.

File: tests/chapter4-exercise-paragraphs-tests.md [hunk 10]
```diff
@@ -48,3 +53,25 @@ Any tags that are open when encountering the second paragraph should be closed
...
+     <html>
+       <body>
+         <p class="first-paragraph">
+           'This has a class'
+         <p id="second-paragraph">
+           'This has an ID'
+		 
+		 
+ Test that parent pointers are set correctly	 
+	>>> test_parse("<div><p>Paragraph within a div</p><b>Bold text</b></div>")
```
Comment (pavpanchekha) [inline tests/chapter4-exercise-paragraphs-tests.md:69]: Literally does not test the parent pointers

File: tests/chapter4-exercise-scripts-tests.md [hunk 11]
```diff
@@ -66,3 +65,17 @@ The script should end only with a complete end script tag.
...
            ' "</scri" "pt>" '
+		   
+		   
+ Test <script> tags with attributes 		   
+    >>> test_parse("<script defer src='my-script.js'>Should be treated as text</script>")
+     <html>
+       <head>
+         <script defer src="my-script.js">
+           'Should be treated as text'
+         </script>
```
Comment (pavpanchekha) [inline tests/chapter4-exercise-scripts-tests.md:76]: No close tags, c'mon dude test this shit

**Review by pavpanchekha — CHANGES_REQUESTED** [review]

File: tests/chapter4-base-tests.md (hunk 3 above)
Comment (pavpanchekha) [reply tests/chapter4-base-tests.md:103]: This isn't even the test! The issue is that some students, in `close_tag`, do:

```
def close_tag(self, node):
    # ...
    if node.tag == "/i":
      # ...
```

However, `node.tag` never includes the slash, they should write `node.tag == "i"`. If they make this mistake, the `</i>` behavior is never triggered and text never stops being italics. *That's* what we want to test for: not parsing, but whether we actually stop italicizing stuff when we reach a `</i>` tag.

**Review by pavpanchekha — CHANGES_REQUESTED** [review]

File: ch03/base.md [hunk 12]
```diff
@@ -79,6 +81,15 @@ rather a mock font that has faked metrics.
...
     [(13.0, 22.5, 'abc', Font size=24 weight=normal slant=italic style=None), 
      (109.0, 25.5, 'def', Font size=20 weight=normal slant=roman style=None)]
+     
+    >>> content = "hello world"
+    >>> url = browser.URL(wbemocks.socket.serve(content))
+    >>> this_browser = browser.Browser()
+    >>> this_browser.load(url)
+    >>> def test_layout(content):
+    ...   dl = browser.Layout(browser.lex(content)).display_list
+    ...   wbemocks.print_list(wbemocks.normalize_display_list(dl))
```
Comment (pavpanchekha) [inline ch03/base.md:91]: ??? Why define this function and then not call it?

Comment (pavpanchekha) [reply ch03/base.md:91]: Also this breaks other tests.

File: ch04/base.md [hunk 13]
```diff
@@ -89,6 +89,16 @@ Attributes can be set on tags:
...
 	       'text'
+		   
+Test that the italic </i> tag is parsed correctly:
+
+	>>> test_parse("<i>A</i>B")
+	 <html>
+	   <body>
+             <i>
+	       'A'
+	     'B'
```
Comment (pavpanchekha) [inline ch04/base.md:100]: You have a lot of whitespace after the `'B'` and this test also fails.

File: ch04/ex-comments.md [hunk 14]
```diff
@@ -94,3 +94,31 @@ There are some edge cases to take care of.
...
+         'This is normal text --> following text should also be normal'
+	 
+	 
+Test combinations "nested" cases
+
+    >>> test_parse("This is normal text<!-- <!-- --> following text should also be normal")
+     <html>
+       <body>
+         'This is normal text' 
+         'following text should also be normal'
```
Comment (pavpanchekha) [inline ch04/ex-comments.md:114]: Why is there no space at the beginning of "following"?

File: ch04/ex-comments.md [hunk 15]
```diff
@@ -94,3 +94,31 @@ There are some edge cases to take care of.
...
+
+	 
+ Test the case of <!-- !--> (should be one comment)
+ 
+    >>> test_parse("This is normal text<!-- !--> following text should also be normal")
+     <html>
+       <body>
+         'This is normal text' 
+         'following text should also be normal'
```
Comment (pavpanchekha) [inline ch04/ex-comments.md:123]: Same, missing space

File: ch04/ex-paragraphs.md [hunk 16]
```diff
@@ -54,3 +54,23 @@ Any tags that are open when encountering the second paragraph should be closed
...
+    >>> test_no_nested_paragraphs("<p>This is a paragraph with <p>another nested paragraph inside.</p></p>")
+     <html>
+       <body>
+         <p>This is a paragraph with 
+         <p>another nested paragraph inside.
+
+
+Test that parent pointers are set correctly			 
+
+    >>>test_parse("<p>This is <b>paragraph one</b> with <p>nested paragraph two</p></p>")
```
Comment (pavpanchekha) [inline ch04/ex-paragraphs.md:70]: Lacks blank after `>>>`

File: ch04/ex-scripts.md [hunk 17]
```diff
@@ -65,3 +65,18 @@ The script should end only with a complete end script tag.
...
          <script>
            ' "</scri" "pt>" '
+		   
+		   
+ Test `<script>` tags with attributes 		   
+    
+    >>> test_parse("<script defer src='my-script.js'>Should be treated as text</script>")
+     <html>
+       <head>
+         <script defer src="my-script.js">
```
Comment (pavpanchekha) [inline ch04/ex-scripts.md:75]:
```suggestion
         <script defer="" src="my-script.js">
```

**Review by pavpanchekha — CHANGES_REQUESTED** [review]

**Comment by KhalidTheeb** [comment]
I opened two other pull requests in place of this outdated pull request.
